File: backend/src/bond_monitor/application/trading/trading_state_use_case.py
"""Combined trading plan + advice in a single broker I/O round-trip."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from bond_monitor.application.trading.advise_use_case import AdviseUseCase
from bond_monitor.application.trading.context import TradingContext
from bond_monitor.application.trading.plan_from_broker import build_trading_plan
from bond_monitor.application.trading.types import TradingAdviceResult
from bond_monitor.domain.bonds.models import BondRecord
from bond_monitor.domain.portfolio.planner import PortfolioPlan
from bond_monitor.domain.portfolio.policies import DurationPolicy
from bond_monitor.application.trading import broker
from bond_monitor.infrastructure.tinvest.snapshot_adapter import (
    broker_active_orders_from_infrastructure,
    broker_operations_from_infrastructure,
    broker_snapshot_from_infrastructure,
)
from bond_monitor.application.trading.advise_use_case import operations_from_date


@dataclass(frozen=True)
class TradingStateResult:
    plan: PortfolioPlan
    advice: TradingAdviceResult


class TradingStateUseCase:
    def __init__(self, ctx: TradingContext) -> None:
        self._ctx = ctx
        self._advise = AdviseUseCase(ctx)

    async def get_trading_state(
        self,
        portfolio_id: str,
        universe: list[BondRecord],
        *,
        key_rate: float,
        tax_rate: float,
        today: date,
        duration_policy: DurationPolicy | None = None,
    ) -> TradingStateResult:
        portfolio = await self._ctx.get_trading_portfolio(portfolio_id)
        token = self._ctx.token(portfolio.account_kind)  # type: ignore[arg-type]
        account_id = portfolio.account_id  # type: ignore[assignment]

        snapshot = broker.get_account_snapshot(token, portfolio.account_kind, account_id)  # type: ignore[arg-type]
        operations = broker.get_account_operations(
            token,
            portfolio.account_kind,  # type: ignore[arg-type]
            account_id=account_id,
            from_date=operations_from_date(portfolio, today=today),
        )
        active_orders = broker.get_active_orders(
            token,
            portfolio.account_kind,  # type: ignore[arg-type]
            account_id=account_id,
        )

        broker_snapshot = broker_snapshot_from_infrastructure(snapshot)
        plan = build_trading_plan(
            portfolio,
            broker_snapshot,
            universe,
            key_rate=key_rate,
            tax_rate=tax_rate,
            today=today,
            duration_policy=duration_policy,
        )
        advice = await self._advise.build_advice_result(
            portfolio,
            universe,
            snapshot=snapshot,
            operations=operations,
            active_orders=active_orders,
            key_rate=key_rate,
            tax_rate=tax_rate,
            today=today,
            duration_policy=duration_policy,
        )
        return TradingStateResult(plan=plan, advice=advice)

File: backend/internal/application/trading/trading_state_use_case.go
package trading

import (
	"context"
	"time"

	"github.com/tonatos/bond-monitor/backend/internal/application"
	appportfolio "github.com/tonatos/bond-monitor/backend/internal/application/portfolio"
	"github.com/tonatos/bond-monitor/backend/internal/domain/bonds"
	domainPortfolio "github.com/tonatos/bond-monitor/backend/internal/domain/portfolio"
	"github.com/tonatos/bond-monitor/backend/internal/infrastructure/tinvest"
)

const operationsLookbackDays = 365

// TradingStateUseCase returns combined trading plan + advice.
type TradingStateUseCase struct {
	ctx    *Context
	advise *AdviseUseCase
	broker *BrokerFacade
	plans  *appportfolio.PlanUseCase
}

func NewTradingStateUseCase(ctx *Context, advise *AdviseUseCase, broker *BrokerFacade, plans *appportfolio.PlanUseCase) *TradingStateUseCase {
	return &TradingStateUseCase{ctx: ctx, advise: advise, broker: broker, plans: plans}
}

func (u *TradingStateUseCase) GetTradingState(ctx context.Context, portfolioID string, universe []bonds.BondRecord, keyRate, taxRate float64, today time.Time, durationPolicy *domainPortfolio.DurationPolicy) (application.TradingStateResult, error) {
	p, err := u.ctx.GetTradingPortfolio(ctx, portfolioID)
	if err != nil {
		return application.TradingStateResult{}, mapTradingErr(err)
	}
	kind := *p.AccountKind
	accountID := *p.AccountID
	snapshot, err := u.broker.GetAccountSnapshot(ctx, kind, accountID)
	if err != nil {
		return application.TradingStateResult{}, err
	}
	ops, err := u.broker.GetAccountOperations(ctx, kind, accountID, OperationsFromDate(today))
	if err != nil {
		return application.TradingStateResult{}, err
	}
	orders, err := u.broker.GetActiveOrders(ctx, kind, accountID)
	if err != nil {
		return application.TradingStateResult{}, err
	}
	brokerSnapshot := tinvest.ToBrokerSnapshot(snapshot)
	brokerOps := tinvest.ToBrokerOperations(ops)
	policy := durationPolicyOrDefault(p, durationPolicy)
	plan := u.plans.BuildForTrading(p, brokerSnapshot, brokerOps, universe, today, keyRate, taxRate, policy)
	advice, err := u.advise.BuildAdviceResult(ctx, p, universe, snapshot, ops, orders, keyRate, taxRate, today, durationPolicy, plan.ResolvedSlots)
	if err != nil {
		return application.TradingStateResult{}, err
	}
	return application.TradingStateResult{Plan: plan, Advice: advice}, nil
}

// OperationsFromDate returns the lower bound for broker operations history.
func OperationsFromDate(today time.Time) time.Time {
	return today.AddDate(0, 0, -operationsLookbackDays)
}

func durationPolicyOrDefault(p domainPortfolio.Portfolio, override *domainPortfolio.DurationPolicy) domainPortfolio.DurationPolicy {
	if override != nil {
		return *override
	}
	return domainPortfolio.DurationPolicyForPortfolio(p, domainPortfolio.RateScenarioHold)
}
